test/benchmarks/benchmark_png_decode.cc

```diff
@@ -0,0 +1,123 @@
+#include <stdio.h>
+#include <benchmark/benchmark.h>
+#include "png_funcs.h"
+#include <assert.h>
+
+class png_decode: public benchmark::Fixture {
+protected:
+    png_dat inpng[10];
+
+    /* Backing this on the heap is a more realistic benchmark */
+    uint8_t *output_img_buf = NULL;
+
+public:
+    /* Let's make the vanilla version have something extremely
+     * compressible */
+    virtual void init_img(png_bytep img_bytes, size_t width, size_t height) {
+        init_compressible(img_bytes, width*height);
+    }
+
+    void SetUp(const ::benchmark::State& state) {
+        output_img_buf = (uint8_t*)malloc(IMWIDTH * IMHEIGHT * 3);
+        assert(output_img_buf != NULL);
+        init_img(output_img_buf, IMWIDTH, IMHEIGHT);
+
+        /* First we need to author the png bytes to be decoded */
+        for (int i = 0; i < 10; ++i) {
+            inpng[i] = {NULL, 0, 0};
+            encode_png(output_img_buf, &inpng[i], i, IMWIDTH, IMHEIGHT);
+        }
+    }
+
+    /* State in this circumstance will convey the compression level */
+    void Bench(benchmark::State &state) {
+        for (auto _ : state) {
+            int compress_lvl = state.range(0);
+            png_parse_dat in = { inpng[compress_lvl].buf };
+            uint32_t width, height;
+            decode_png(&in, (png_bytepp)&output_img_buf, IMWIDTH * IMHEIGHT * 3, width, height);
+        }
+    }
+
+    void TearDown(const ::benchmark::State &state) {
+        free(output_img_buf);
+        for (int i = 0; i < 10; ++i) {
+            free(inpng[i].buf);
+        }
+    }
+};
+
+class png_decode_realistic: public png_decode {
+private:
+    bool test_files_found = false;
+
+public:
+    void SetUp(const ::benchmark::State &state) {
+        output_img_buf = NULL;
+        output_img_buf = (uint8_t*)malloc(IMWIDTH * IMHEIGHT * 3);
+        /* Let's take all the images at different compression levels and jam their bytes into buffers */
+        char test_fname[25];
+        FILE *files[10];
+
+        /* Set all to NULL */
+        memset(files, 0, sizeof(FILE*));
+        for (size_t i = 0; i < 10; ++i)
+            inpng[i] = { NULL, 0, 0 };
+
+        for (size_t i = 0; i < 10; ++i) {
```

Comment:
This for loop and the one where you are reading the bytes in the file, I think they can just be merged into one for loop.  

Actually, all these for loops can be merged into one. I think it would be cleaner.

Reply:
The reason I did that is because the destructor of the class frees everything.  Because the error handling can happen mid loop, everything needs to be initialized as NULL or a wild pointer will be freed.

Reply:
It can happen but then the cleanup becomes an extra for loop for setting the not yet traversed indices to NULL so that TearDown() doesn't crash

Reply:
Added what was necessary to do this, have a look.

Reply:
Oh for the fopen and fread - I'd rather there be all or nothing on that so that we don't have to test for the existence of a valid buffer to decode a PNG per test.